data_structures/rtree.cpp

```diff
@@ -0,0 +1,168 @@
+// rtree.cpp
+
+#include <iostream>
+#include <memory>
+
+#include "rtree.hpp"
+
+using coord_t = spatial::coord_t;
+using code_t = spatial::code_t;
+using index_t = spatial::index_t;
+
+int const M = 8;
+
+template<typename T>
+spatial::Rtree<T>::Rtree(): 
+    root_entry(std::make_shared<InternalEntry>()) 
```

Comment:
This is maybe a confusing part of the r-tree, but you always build the tree upwards, not downwards. Thus, your initial entry should be a `LeafEntry`

Reply:
I think this is poor documentation on my part, but the root node does start off as a leaf (as in it contains a number of `LeafEntry` structs). I'm wrapping up the root node in an `InternalEntry` just so that I can keep track of the bounding box of the root.